### ISSUE TITLE
regression tests for mitmproxy

### ISSUE THREAD

**Issue by msridhar** [Samsung/jalangi2#124]
We need regression tests for our mitmproxy support that run on Travis.  Ideally this support would also test the certificate installation process.  Not sure if this is possible.

**Comment by csujedihy** [comment]
mitmproxy crash when accessing mtalk.google.com. Just a heads up.

**Comment by msridhar** [comment]
@csujedihy can you try with the latest installation instructions in the README?  In particular, you should use mitmproxy version 0.11.3.  Before trying, please do `pip uninstall mitmproxy` and `rm -rf ~/.mitmproxy`.  You should also remove any mitmproxy certificates you placed in the Keychain (if on Mac, otherwise the appropriate Linux thing) and reinstall freshly-generated certificates.

**Comment by csujedihy** [comment]
My mitmdump's version is 0.18.2.

**Comment by csujedihy** [comment]
I got it. You modified the README recently.

**Comment by msridhar** [comment]
Yes, just this morning 😄 mitmproxy 0.17 *might* work, 0.18.2 definitely doesn't.  This is the motivation for this issue

**Comment by msridhar** [comment]
I have confirmed that [jalangi2analyses](https://github.com/ksen007/jalangi2analyses) do *not* work out of the box with mitmproxy 0.17 installed.  This may just be updating which APIs get invoked, but I also see various errors in the console about the client not trusting the proxy's certificate.  So there may be a mitmproxy problem here.

**Comment by msridhar** [comment]
@christofferqa have you gotten mitmproxy 0.17 to work with HTTPS sites on Mac?

**Comment by christofferqa** [comment]
Yes. Have you marked the mitmproxy certificate as trusted in Keychain Access?

**Comment by csujedihy** [comment]
0.17-0.18.2 will crash when visiting mtalk.google.com via HTTPS.

**Comment by msridhar** [comment]
@csujedihy does 0.11.3 work on mtalk.google.com?

**Comment by christofferqa** [comment]
I don't experience the problems you are having with 0.17.1 (https://github.com/mitmproxy/mitmproxy/tree/v0.17.1). My steps were as follows (assuming mitmproxy has been installed properly).

1. Start mitmproxy.
   ```
   mitmdump --quiet --anticache -s "scripts/proxy.py --inlineIID --inlineSource --analysis src/js/sample_analyses/ChainedAnalyses.js --analysis src/js/runtime/analysisCallbackTemplate.js"
   ```

2. Open Google Chrome with the proper proxy flags.

   Mac OS:
   ```
   open -a 'Google Chrome' --args --proxy-server="127.0.0.1:8080" --proxy-bypass-list=""
   ```
   Ubuntu:
   ```
   google-chrome-stable --proxy-server="127.0.0.1:8080" --proxy-bypass-list=""
   ```

3. Open https://hangouts.google.com. (I tried opening mtalk.google.com, but it is not reachable for me.)


Below is a screenshot of Jalangi2 running on Google Hangouts (on HTTPS). (Uploaded to http://users-cs.au.dk/cqa/jalangi2/hangouts.png.)

![Jalangi2 on Google Hangouts](http://users-cs.au.dk/cqa/jalangi2/hangouts.png)

Below is a screenshot of Google Hangouts (on HTTPS), when visited using mitmproxy 0.17.1, but *without* the Jalangi2 instrumentation. (Uploaded to http://users-cs.au.dk/cqa/jalangi2/hangouts-no-instr.png.)

![Jalangi2 on Google Hangouts](http://users-cs.au.dk/cqa/jalangi2/hangouts-no-instr.png)

**Comment by msridhar** [comment]
Christoffer, can you clarify the exact installation steps you use for
mitmproxy?

**Comment by christofferqa** [comment]
Yes, but I'll need access to a Mac, where mitmproxy has not previously been installed, to be sure that I don't have some hidden dependency installed. It may take a few days.

**Comment by msridhar** [comment]
When you do get to a Mac, let us know if it's running El Capitan or
Sierra.  My issues have been on Sierra; maybe that's the problem.

**Comment by christofferqa** [comment]
**Complete jalangi2 install instructions for OS X 10.10.** *Tested on an entirely fresh virtual machine, installed from the recovery partition of my Mac (see http://cs.au.dk/~cqa/jalangi2/vm.png).*

1) Install Homebrew

```
$ /usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```

2) Install git, node 4.x, python

```
$ brew install git
$ brew install node@4
$ brew install python
```

3) Update pip and install mitmproxy 0.17

```
$ pip install --upgrade pip
$ pip install mitmproxy==0.17
```

4) Install Google Chrome from https://www.google.com/chrome

5) Install mitmproxy certificates

   A. Start mitmproxy and open http://mitm.it/ in Chrome, with the proper proxy flags (make sure not to have a running instance of Chrome!)

   ```
   $ mitmdump -p 8081
   $ open -a 'Google Chrome' --args http://mitm.it/ --proxy-server="127.0.0.1:8081" --proxy-bypass-list=""
   ```

   B. Click on the Apple icon in order to download "mitmproxy-ca-cert.pem", and then open "mitmproxy-ca-cert.pem" to install the certificate

   C. Open Keychain Access, go to the "Certificates" menu, right-click on the "mitmproxy" certificate and select "Get Info". In the window that opens, click on the "Trust" item, and set "When using this certificate:" to "Always Trust".

6) Clone jalangi2 and install dependencies

```
$ git clone https://github.com/Samsung/jalangi2.git
$ (cd jalangi2; npm install)
```

7) Start jalangi2

```
$ cd jalangi2
$ mitmdump -p 8081 --quiet --anticache -s "scripts/proxy.py --inlineIID --inlineSource --analysis src/js/sample_analyses/ChainedAnalyses.js --analysis src/js/runtime/analysisCallbackTemplate.js"
```

8) Open Chrome with the proper proxy flags (make sure not to have a running instance of Chrome!), and visit your favorite HTTPS website

```
$ open -a 'Google Chrome' --args --proxy-server="127.0.0.1:8081" --proxy-bypass-list=""
```

**Comment by msridhar** [comment]
FWIW, mitmproxy is now known to work using the master branch, and we've updated the relevant documentation.  We still need regression tests; renamed the issue accordingly
